lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -576,10 +578,9 @@ def __init__(
             raise MissingBulkSchemaParameters(
                 api_name, "dataset parameter is not defined or not required"
             )
-        if self.schemas[ApiMethod.POST].authorization != ApiAuthorizationType.DATASET:
-            raise MissingBulkSchemaParameters(
-                api_name, "schema authorization is not by dataset"
-            )
+        assert (
+            self.schemas[ApiMethod.POST].authorization == ApiAuthorizationType.DATASET
+        ), f"API {self.__class__.__name__} authorization type must be DATASET"
```

Comment:
I've lost my handle on this...why is this an assertion when the preceding checks in the existing code use `if` statements?  Are we anticipating some other yet-to-land PR which will converts the `if` statements to `assert`'s?

I feel like it would be more appropriate for _this_ PR to use an `if` and to let the subsequent PR (if and when it lands) to convert the `if` here to an `assert` during its rebase/merge.

Reply:
One of Peter's PRs started converting these sort of constructor tests into asserts; Siddardh had earlier changed this one to allow multiple authorization types, and converted it to an `assert` at the same time. When he dropped the extra term he left it an `assert`. I suppose arguably the entire statement should just have been reverted at that point... I didn't feel like worrying about it.

On the other hand, having been paying more attention to `PUT` for various reasons recently, I've started to become aware that I had a fundamentally flawed understanding of Flask behavior. I'd always thought that the UI endpoint class was instantiated *once* at initialization and then the `put`/`post`/etc. methods were called on that object. I no longer believe this to be true, because the `upload` `__init__` has a logger message that is appearing at each `PUT` call. I think they're created dynamically, in which case I'm not sure I actually understand what the `assert` will do. Though the simple `raise` probably isn't what we want, either. I'd been hoping that all the `assert` statements would be interpreted and fire *once* during startup and blow up the world... I'm not thrilled with simply having each API call fail with an internal server error (which the `raise` will definitely do), but I don't really understand what `Flask` will do with the `assert`, which we're not handling, and whether this simply leads to bizarre chaos. I've also been carefully avoiding thinking too much about the abyss...

Reply:
>I've lost my handle on this...why is this an assertion when the preceding checks in the existing code use if statements? Are we anticipating some other yet-to-land PR which will converts the if statements to assert's?



Yeah @webbnh, There is one merged [PR](https://github.com/distributed-system-analysis/pbench/commit/c323969164db77aea317f86bcc4f61bd5d5e0b6c)  where `if` are converted to `assert`.

Before merging the `publish` and `update`  API, there is a constraint for ADMIN operation for `update`, so this condition needs to be updated. And Dave suggested using `assert` to maintain uniformity.

Reply:
> Dave suggested using assert to maintain uniformity.

🤣 "Uniformity" with some _other_ module (or class)??

From the sounds of Dave's reply above, we should be moving all of these checks to a unit test, so that they are performed once at _build-time_ and so that they will not burden us at run-time....

Reply:
In fact, for all of the Elasticsearch APIs, `commons.py` already constructs an instance of each subclass to check the schemas. They should already be failing if any of these `raise` or `assert` things fire. I'm not so sure about trying to *apply* the basic checks solely within the unit test. These constraints are applied *in* a base constructor and need to be *validated against* all subclasses; whereas I'd much prefer that subclass-specific unit tests didn't need to duplicate all this or even be aware of them, and if we had a superclass test (e.g., for `ElasticBase` *it* shouldn't need to be aware of the subclasses that might actually violate the constraints. And it's too much special casing for the already overly general `commons.py`.

I've often thought we need a more complete `commons` hierarchy for `ApiBase` -> `ElasticBase` / `ElasticBulkBase` -> `IndexMapBase` rather than trying to do everything in one, and that would eliminate a lot of the annoying skips. Then the `TestDatasetsUpdate` test could transparently leverage `TestElasticBulkBase` superclass checks for stuff like this without needing to be explicitly aware of them.

In fact ... that would probably be a good story to have under the unit test improvements. I think I'll go make one.